### PR TITLE
Rotations in sorted array

### LINKED ISSUE #1634: No. of times Sorted Array is Rotated in JAVA
This will be done in O(Log n) Time Complexity using Binary Search.

### PR DESCRIPTION
**Pull request by ishaan2908**
Fixes #1634 

#### Checklist:
- [x] 4 space indentation.
- [x] Coding conventions are followed.
- [x] Input is taken dynamically.
- [x] Sample Input / Output is added at the end of file.
- [x] Logic Documentation (Comments).
- [x] File names are correct.

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

#### Changes proposed in this pull request:
- Added method to find how many times a sorted array is rotated using Binary Search

#### Languages Used:
- JAVA

#### Files Added:
- RotationCount.java

> We're happy to help you get this ready -- don't be afraid to ask for help.

Thanks!

### CODE REVIEW

File: Binary_Search/RotationCount.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
+import java.util.Scanner;
+
+class RotationCount
+{
+
+	static int findRotations(int[] arr)
```
Comment (SakshiShreya) [inline Binary_Search/RotationCount.java]: Please use 4 spaces instead of 1 tab for indentation.

Comment (ishaan2908) [reply Binary_Search/RotationCount.java]: Sure! I will make the changes.

**Comment by ishaan2908** [comment]
Can you please review? @SakshiShreya

File: Binary_Search/RotationCount.java [hunk 2]
```diff
@@ -3,69 +3,69 @@
...
-			else if(arr[middle] >= arr[left])
-				left = middle + 1;
-		}
-		return 0;
+        while(left <= right)
+        {
+            int middle = (right + left)/2;
+            int previous = (middle+size-1) % size, next = (middle+1) % size;
+            
+            // variables previous and next store the array elements before and after the middle element respectively.
```
Comment (SakshiShreya) [inline Binary_Search/RotationCount.java]: One line can contain only 80 characters.

File: Binary_Search/RotationCount.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
+import java.util.Scanner;
```
Comment (SakshiShreya) [inline Binary_Search/RotationCount.java]: Please add the description of the algo at the beginning of the file.

**Comment by ishaan2908** [comment]
I have made the requested changes! @SakshiShreya

File: Binary_Search/RotationCount.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
+/*
```
Comment (jainaman224) [inline Binary_Search/RotationCount.java]: Incorrect folder name.

Comment (ishaan2908) [reply Binary_Search/RotationCount.java]: Should I, then, put it under Array_Rotation or create a new folder? @jainaman224

**Comment by ishaan2908** [comment]
I have changed the folder name @jainaman224 . Please review.

File: countRotations/countRotations.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        // count variable stores no. of rotations in the array.    
+        
+        System.out.println("The array is rotated " + count + " times.");
+
+    }
+
+}
+
+    /*
+    *    Sample Input
```
Comment (jainaman224) [inline countRotations/countRotations.java]: Unindent this comment I/O section.

File: countRotations/countRotations.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
+/*
```
Comment (jainaman224) [inline countRotations/countRotations.java]: Change file and folder name to `Count_Rotations`

**Comment by ishaan2908** [comment]
Done! @jainaman224

File: Count_Rotations/Count_Rotations.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+        /*
+        * returns no. of times a sorted array was rotated.\
+        */
+        int size = arr.length;
+        int left = 0, right = size - 1;
+
+        while(left <= right)
+        {
+            int middle = (right + left)/2;
+            int previous = (middle+size-1) % size, next = (middle+1) % size;
```
Comment (somya-kapoor) [inline Count_Rotations/Count_Rotations.java]: Add space around the operator

File: Count_Rotations/Count_Rotations.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    {    
+        /*
+        * returns no. of times a sorted array was rotated.\
+        */
+        int size = arr.length;
+        int left = 0, right = size - 1;
+
+        while(left <= right)
+        {
+            int middle = (right + left)/2;
```
Comment (somya-kapoor) [inline Count_Rotations/Count_Rotations.java]: Add space around the operator(/)

File: Count_Rotations/Count_Rotations.java [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+
+    }
+
+}
+
+/*
+*    Sample Input
+*    size = 6
+*    arr = 5, 6, 1, 2, 3, 4
+*
```
Comment (somya-kapoor) [inline Count_Rotations/Count_Rotations.java]: remove the line

File: Count_Rotations/Count_Rotations.java [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+        for(int i = 0; i < size; i++)
+            arr[i] = scan.nextInt();
+
+        int count = findRotations(arr);
+        // count variable stores no. of rotations in the array.    
+        
+        System.out.println("The array is rotated " + count + " times.");
+
+    }
+
```
Comment (somya-kapoor) [inline Count_Rotations/Count_Rotations.java]: remove the line

File: Count_Rotations/Count_Rotations.java [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+        System.out.println("Enter " + size + " elements: ");
+        for(int i = 0; i < size; i++)
+            arr[i] = scan.nextInt();
+
+        int count = findRotations(arr);
+        // count variable stores no. of rotations in the array.    
+        
+        System.out.println("The array is rotated " + count + " times.");
+
```
Comment (somya-kapoor) [inline Count_Rotations/Count_Rotations.java]: remove the line

File: Count_Rotations/Count_Rotations.java [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+            else if(arr[middle] >= arr[left])
+                left = middle + 1;
+        }
+        return 0;
+
+    }
+    
+    public static void main(String[] args)
+    {
+
```
Comment (somya-kapoor) [inline Count_Rotations/Count_Rotations.java]: remove the line

File: Count_Rotations/Count_Rotations.java [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+                return middle;
+            
+            if(arr[middle] <= arr[right])
+                right = middle - 1;
+            
+            else if(arr[middle] >= arr[left])
+                left = middle + 1;
+        }
+        return 0;
+
```
Comment (somya-kapoor) [inline Count_Rotations/Count_Rotations.java]: remove the line

File: Count_Rotations/Count_Rotations.java [hunk 14]
```diff
@@ -0,0 +1,77 @@
+/*
+*   A sorted array has been rotated in clockwise direction.   
+*   Find how many times it was rotated.
+*/
+
+import java.util.Scanner;
+
+class CountRotations
+{
+
```
Comment (somya-kapoor) [inline Count_Rotations/Count_Rotations.java]: remove the line

**Comment by ishaan2908** [comment]
Done! @somya-kapoor

**Comment by somya-kapoor** [comment]
Please squash the commits

**Comment by somya-kapoor** [comment]
@ishaan2908 Are you working on it?

**Comment by ishaan2908** [comment]
Yes @somya-kapoor, but facing issues in squashing commits. Can you please use 'Squash and Merge' ?

**Comment by ishaan2908** [comment]
Done! I have squashed the commits @somya-kapoor

**Pull request merged**